### PR TITLE
feat: [CI-4307]: Enable dockerhub authentication in ECR plugin

### PR DESCRIPTION
**Pull request by raghavharness**
You can use the following comments to re-trigger PR Checks

- Compile: `trigger compile`
- runAeriformCheck: `trigger AeriformCheck`
- CodeFormat: `trigger codeformat`
- MessageMetadata: `trigger messagecheck`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- runDockerizationCheck: `trigger dockerizationcheck`
- runAuthorCheck: `trigger authorcheck`
- Checkstyle: `trigger checkstyle`
- PMD: `trigger pmd`
- TI-bootstrap: `trigger ti0`
- TI-bootstrap1: `trigger ti1`
- TI-bootstrap2: `trigger ti2`
- TI-bootstrap3: `trigger ti3`
- TI-bootstrap4: `trigger ti4`
- FunctionalTest1: `trigger ft1`
- FunctionalTest2: `trigger ft2`
- CodeBaseHash: `trigger codebasehash`

---
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/32917)

### CODE REVIEW

File: 330-ci-beans/src/main/java/io/harness/beans/steps/stepinfo/ECRStepInfo.java [hunk 1]
```diff
@@ -94,16 +97,21 @@ public class ECRStepInfo implements PluginCompatibleStep, WithConnectorRef {
   @YamlSchemaTypes({string}) @ApiModelProperty(dataType = INTEGER_CLASSPATH) private ParameterField<Integer> runAsUser;
   @YamlSchemaTypes({string}) @ApiModelProperty(dataType = BOOLEAN_CLASSPATH) private ParameterField<Boolean> optimize;
   @ApiModelProperty(dataType = STRING_CLASSPATH) private ParameterField<String> remoteCacheImage;
+  @YamlSchemaTypes(value = {string})
+  @ApiModelProperty(dataType = STRING_MAP_CLASSPATH)
+  private ParameterField<Map<String, String>> baseImageConnectorRefs;
```
Comment (shubham149) [inline 330-ci-beans/src/main/java/io/harness/beans/steps/stepinfo/ECRStepInfo.java:102]: Shouldn't this be a list? What's the key and value for this map?

Comment (raghavharness) [reply 330-ci-beans/src/main/java/io/harness/beans/steps/stepinfo/ECRStepInfo.java:102]: Based on discussion with Vardan and Rajatha, since UI doesn't support runtime Inputs on list fields therefore it was decided to be kept as a map. Key is not really useful here, the value is the actual connectorRef.

File: 320-ci-execution/src/main/java/io/harness/ci/integrationstage/VmInitializeStepUtils.java [hunk 2]
```diff
@@ -93,7 +95,7 @@ public BuildJobEnvInfo getInitializeStepInfoBuilder(StageElementConfig stageElem
       }
     }
     IntegrationStageConfig integrationStageConfig = IntegrationStageUtils.getIntegrationStageConfig(stageElementConfig);
-    validateStageConfig(integrationStageConfig, accountId);
+    validateStageConfig(integrationStageConfig, AmbianceUtils.getAccountId(ambiance));
```
Comment (shubham149) [inline 320-ci-execution/src/main/java/io/harness/ci/integrationstage/VmInitializeStepUtils.java:98]: VMs uses a different plugin i.e. plugins/ecr for pushing images to ecr. Currently it won't work with base image connector ref on vm builds.

File: 330-ci-beans/src/main/java/io/harness/beans/steps/stepinfo/ECRStepInfo.java (hunk 1 above)
Comment (shubham149) [reply 330-ci-beans/src/main/java/io/harness/beans/steps/stepinfo/ECRStepInfo.java:102]: It should be a list only. This cannot be a map. We cannot have misleading api for the UI to work. There are lot of attributes that are already list for a couple of steps.

File: 320-ci-execution/src/main/java/io/harness/ci/integrationstage/VmInitializeStepUtils.java (hunk 2 above)
Comment (raghavharness) [reply 320-ci-execution/src/main/java/io/harness/ci/integrationstage/VmInitializeStepUtils.java:100]: Added exception for VM Infra

File: 330-ci-beans/src/main/java/io/harness/beans/steps/stepinfo/ECRStepInfo.java (hunk 1 above)
Comment (raghavharness) [reply 330-ci-beans/src/main/java/io/harness/beans/steps/stepinfo/ECRStepInfo.java]: Changed it to list.

File: 860-orchestration-steps/src/main/java/io/harness/filters/GenericStepPMSFilterJsonCreatorV2.java [hunk 3]
```diff
@@ -58,6 +58,9 @@ public FilterCreationResponse handleNode(FilterCreationContext filterCreationCon
       String projectIdentifier = filterCreationContext.getSetupMetadata().getProjectId();
       Map<String, ParameterField<String>> connectorRefs =
           ((WithConnectorRef) yamlField.getStepSpecType()).extractConnectorRefs();
+      if (yamlField.getStepSpecType() instanceof WithBaseImageConnectorRef) {
```
Comment (harshjain12) [inline 860-orchestration-steps/src/main/java/io/harness/filters/GenericStepPMSFilterJsonCreatorV2.java:61]: No need to change this code

File: 320-ci-execution/src/main/java/io/harness/stateutils/buildstate/PluginSettingUtils.java [hunk 4]
```diff
@@ -131,9 +132,19 @@ public String getConnectorRef(PluginCompatibleStep stepInfo) {
...
 
-  public Map<EnvVariableEnum, String> getConnectorSecretEnvMap(CIStepInfo stepInfo) {
+  public List<String> getBaseImageConnectorRefs(PluginCompatibleStep stepInfo) {
+    if (stepInfo instanceof WithBaseImageConnectorRef) {
+      WithBaseImageConnectorRef withBaseImageStepInfo = (WithBaseImageConnectorRef) stepInfo;
+      String stepType = stepInfo.getNonYamlInfo().getStepInfoType().getDisplayName();
+      return RunTimeInputHandler.resolveListParameter("baseImageConnectorRefs", stepType, stepInfo.getIdentifier(),
+          withBaseImageStepInfo.getBaseImageConnectorRefs(), false);
+    }
+    return null;
```
Comment (harshjain12) [inline 320-ci-execution/src/main/java/io/harness/stateutils/buildstate/PluginSettingUtils.java:142]: Always return empty list

File: 330-ci-beans/src/main/java/io/harness/beans/plugin/compatible/PluginCompatibleStep.java [hunk 5]
```diff
@@ -14,10 +14,18 @@
...
 public interface PluginCompatibleStep extends CIStepInfo {
   // Common for all plugin compatible step types
   ParameterField<String> getConnectorRef();
   ContainerResource getResources();
   ParameterField<Integer> getRunAsUser();
+  default boolean supportsBaseImageConnectorRef() {
+    return false;
+  }
+  default ParameterField<List<String>> getBaseImageConnectorRefs() {
+    return null;
```
Comment (harshjain12) [inline 330-ci-beans/src/main/java/io/harness/beans/plugin/compatible/PluginCompatibleStep.java:29]: Do not return null here
Return empty parameter field

File: 878-ng-common-utilities/src/main/java/io/harness/PipelineSetupUsageUtils.java [hunk 6]
```diff
@@ -52,22 +55,36 @@ public List<EntityDetail> extractInputReferredEntityFromYaml(String accountIdent
       if (!metadata.containsKey(PreFlightCheckMetadata.EXPRESSION)) {
         entityDetails.add(referredUsage.getReferredEntity());
       } else if (fqnToObjectMapMergedYaml.containsKey(fqn)) {
-        String finalValue = ((TextNode) fqnToObjectMapMergedYaml.get(fqn)).asText();
```
Comment (harshjain12) [inline 878-ng-common-utilities/src/main/java/io/harness/PipelineSetupUsageUtils.java:55]: Are these changes still required ?

File: 330-ci-beans/src/main/java/io/harness/beans/plugin/compatible/PluginCompatibleStep.java (hunk 5 above)
Comment (raghavharness) [reply 330-ci-beans/src/main/java/io/harness/beans/plugin/compatible/PluginCompatibleStep.java:29]: okay

File: 320-ci-execution/src/main/java/io/harness/stateutils/buildstate/PluginSettingUtils.java [hunk 7]
```diff
@@ -131,9 +132,18 @@ public String getConnectorRef(PluginCompatibleStep stepInfo) {
         "connectorRef", stepType, stepInfo.getIdentifier(), stepInfo.getConnectorRef(), true);
   }
 
-  public Map<EnvVariableEnum, String> getConnectorSecretEnvMap(CIStepInfo stepInfo) {
+  public List<String> getBaseImageConnectorRefs(PluginCompatibleStep stepInfo) {
+    if (stepInfo.supportsBaseImageConnectorRef()) {
```
[the remark this answers is not in the document]
Comment (raghavharness) [reply 320-ci-execution/src/main/java/io/harness/stateutils/buildstate/PluginSettingUtils.java:136]: What will be null ?

File: 878-ng-common-utilities/src/main/java/io/harness/PipelineSetupUsageUtils.java (hunk 6 above)
Comment (raghavharness) [reply 878-ng-common-utilities/src/main/java/io/harness/PipelineSetupUsageUtils.java:55]: yeah these changes are required for supporting baseImageConnectorRefs as runtimeInput.

File: 320-ci-execution/src/main/java/io/harness/stateutils/buildstate/PluginSettingUtils.java (hunk 4 above)
Comment (raghavharness) [reply 320-ci-execution/src/main/java/io/harness/stateutils/buildstate/PluginSettingUtils.java]: done

File: 860-orchestration-steps/src/main/java/io/harness/filters/GenericStepPMSFilterJsonCreatorV2.java (hunk 3 above)
Comment (raghavharness) [reply 860-orchestration-steps/src/main/java/io/harness/filters/GenericStepPMSFilterJsonCreatorV2.java]: done

File: 320-ci-execution/src/main/java/io/harness/ci/integrationstage/K8InitializeStepInfoBuilder.java [hunk 8]
```diff
@@ -559,49 +566,81 @@ private StepElementConfig getStepElementConfig(ExecutionWrapperConfig executionW
...
     }
     return map;
   }
 
+  private List<K8BuildJobEnvInfo.ConnectorConversionInfo> getBaseImageConnectorConversionInfo(
+      PluginCompatibleStep step, Ambiance ambiance) {
+    List<String> baseConnectorRefs = PluginSettingUtils.getBaseImageConnectorRefs(step);
+    List<K8BuildJobEnvInfo.ConnectorConversionInfo> baseImageConnectorConversionInfos = new ArrayList<>();
+    NGAccess ngAccess = AmbianceUtils.getNgAccess(ambiance);
+    if (baseConnectorRefs != null) {
```
Comment (harshjain12) [inline 320-ci-execution/src/main/java/io/harness/ci/integrationstage/K8InitializeStepInfoBuilder.java:622]: check with isEmpty

File: 320-ci-execution/src/main/java/io/harness/ci/integrationstage/K8InitializeStepInfoBuilder.java [hunk 9]
```diff
@@ -559,49 +566,81 @@ private StepElementConfig getStepElementConfig(ExecutionWrapperConfig executionW
...
+      for (String baseConnectorRef : baseConnectorRefs) {
```
Comment (harshjain12) [inline 320-ci-execution/src/main/java/io/harness/ci/integrationstage/K8InitializeStepInfoBuilder.java:623]: use stream iterator

File: 330-ci-beans/src/main/java/io/harness/beans/serializer/RunTimeInputHandler.java [hunk 10]
```diff
@@ -282,6 +282,7 @@ public List<String> resolveListParameter(String fieldName, String stepType, Stri
             format("Failed to resolve mandatory field %s in step type %s with identifier %s", fieldName, stepType,
                 stepIdentifier));
       }
+      return null;
```
Comment (harshjain12) [inline 330-ci-beans/src/main/java/io/harness/beans/serializer/RunTimeInputHandler.java:285]: return empty list

File: 330-ci-beans/src/main/java/io/harness/beans/steps/stepinfo/ECRStepInfo.java [hunk 11]
```diff
@@ -142,6 +150,24 @@ public String getFacilitatorType() {
   public Map<String, ParameterField<String>> extractConnectorRefs() {
     Map<String, ParameterField<String>> connectorRefMap = new HashMap<>();
     connectorRefMap.put(YAMLFieldNameConstants.CONNECTOR_REF, connectorRef);
+    connectorRefMap.putAll(extractBaseImageConnectorRefs());
     return connectorRefMap;
   }
+
+  private Map<String, ParameterField<String>> extractBaseImageConnectorRefs() {
```
Comment (harshjain12) [inline 330-ci-beans/src/main/java/io/harness/beans/steps/stepinfo/ECRStepInfo.java:157]: move this to utility.

File: 330-ci-beans/src/main/java/io/harness/beans/serializer/RunTimeInputHandler.java (hunk 10 above)
Comment (raghavharness) [reply 330-ci-beans/src/main/java/io/harness/beans/serializer/RunTimeInputHandler.java:285]: Other methods also sends null for such cases in RuntimeInputHandler. This was missed for List so I added null. Should we keep it consistent with other definitions ?

File: 878-ng-common-utilities/src/main/java/io/harness/PipelineSetupUsageUtils.java (hunk 6 above)
Comment (raghavharness) [reply 878-ng-common-utilities/src/main/java/io/harness/PipelineSetupUsageUtils.java:55]: @sahilhindwani-harness Can you please help in reviewing these changes.

**Comment by raghavharness** [comment]
trigger ti0
